### PR TITLE
Fixes #9504: Edit logins

### LINKED ISSUE #9504: Edit Logins
## User Story
As a user, I want to be able to edit my logins, so I have control over changing them manually. 

### Acceptance Criteria
I can edit my logins in Logins settings
   - [x] Add kebab (three dot) menu to login info toolbar with options to edit and delete
   - [x] Add fragment with editable fields
     * disable web address editing - YES
   - [ ] Error states
     * possibly break this out into a separate piece of work?
   - [x] Save enabled only when valid
     * Successful save routes to site info page
   - [ ] Cancel (`X`) button 
     * where does cancel route to - login info
     * do we want a confirmation dialog - no
   - [x] Feature flags

## Designs
[UX Mocks](https://share.goabstract.com/33afd8de-f90e-4504-90e3-ee2a3bcb1925?mode=design&sha=2b7a1bf44fa2c02a3b7e470075e94c587a491d3b
)

### SVG Icons
[checkmark](https://drive.google.com/open?id=16GAx-27bja-hxaKpgaRVNrzcsiS1gJ1p) 
[drop-down arrow](https://drive.google.com/open?id=16evm0KYF898lgt9UKqw8l-rrBFX9Z3gD) 
[warning](https://drive.google.com/open?id=182eby9q2g-vviWa1LfHcDXluQzUEIdmn) 
[clear text field ](https://drive.google.com/open?id=127i_O5i1wfREI170426Pz6WSQGh3heMe)

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #9504 

## Testing notes
The edit functionality is behind a feature flag - `FeatureFlags.loginEdit`. In order to manually test, set the flags to `true`. This will allow you to see the three-dot kebab menu in the login detail screen and enter edit mode.

### Todo
   - [x] Add kebab (three dot) menu to login info toolbar with options to edit and delete
   - [x] Add fragment with editable fields
     * disable web address editing - YES
   - [ ] ~Error states~
     * break this out into a separate piece of work: https://github.com/mozilla-mobile/fenix/issues/10173 and https://github.com/mozilla-mobile/fenix/issues/10172
     * closing keyboards at the right times and clickable state
   - [x] Save enabled only when valid
     * Successful save routes to site info page
   - [x] Cancel (`X`) button 
     * where does cancel route to - login info
     * do we want a confirmation dialog - no
   - [x] Feature flags

## Screenshots
**Consistency between detail view and edit view:**
ux_consistency_edit_login

**Focused field:**
image

**Dark mode**
image image

**Private mode**
image image

----- 

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by eliserichards** [comment]
@ekager I still need to do linting on this, but let's get together and I'll walk you through the PR (sorry it's obnoxiously big)

File: app/src/main/java/org/mozilla/fenix/FeatureFlags.kt [hunk 1]
```diff
@@ -43,4 +46,18 @@ object FeatureFlags {
      * Enables picture-in-picture feature
      */
     val pictureInPicture = Config.channel.isNightlyOrDebug
+
+    /**
+     * Allows edit of saved logins.
+     */
+    val loginEdit = when {
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/FeatureFlags.kt:53]: I might be misunderstanding but why can't we just use the convention of `Config.channel.isNightlyOrDebug` here?

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -1293,8 +1295,24 @@
...
     <!-- Login overflow menu delete button -->
     <string name="login_menu_delete_button">Delete</string>
+    <!-- Login overflow menu edit button -->
+    <string name="login_menu_edit_button">Edit</string>
     <!-- Message in delete confirmation dialog for logins -->
     <string name="login_deletion_confirmation">Are you sure you want to delete this login?</string>
     <!-- Positive action of a dialog asking to delete  -->
     <string name="dialog_delete_positive">Delete</string>
+    <!--  The saved login options menu description. -->
+    <string name="login_menu_more_options">More options</string>
```
Comment (ekager) [inline app/src/main/res/values/strings.xml:1305]: Let's change this to something like "Logins Menu" to be more contextual

File: app/src/main/res/menu/login_options_menu.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/loginOptionsEditDelete"
+    android:contentDescription="@string/login_menu_more_options"
```
Comment (ekager) [inline app/src/main/res/menu/login_options_menu.xml:10]: I'm not sure we need to match this theme since this 3 dot menu will now look different from every other menu that we have in the app. Let's double check with UX about this.

File: app/src/main/res/menu/login_delete.xml [hunk 4]
```diff
@@ -1,14 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
+
 <!-- This Source Code Form is subject to the terms of the Mozilla Public
    - License, v. 2.0. If a copy of the MPL was not distributed with this
    - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+
 <menu xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (ekager) [inline app/src/main/res/menu/login_delete.xml:7]: Can we delete this whole file? I don't think we're using it anymore?

File: app/src/main/res/menu/login_options_menu.xml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    android:theme="@style/PopupKebabMenu"
+    android:title="@string/login_menu_more_options"
+    app:showAsAction="never" >
+
+    <item
+        android:id="@+id/edit_login_button"
+        android:contentDescription="@string/login_menu_delete_button"
+        android:icon="@drawable/ic_edit"
+        android:minWidth="82dp"
+        android:padding="20dp"
```
Comment (ekager) [inline app/src/main/res/menu/login_options_menu.xml:21]: Same comment as the top of this file, I think customizing the items in the menu are just going to make this menu inconsistent with every other menu in the app

File: app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            try {
+                if (!passwordText.text.isNullOrBlank()) {
+                    attemptSaveAndExit()
+                } else {
+                    view?.let {
+                        FenixSnackbar.make(
+                            view = it,
+                            duration = Snackbar.LENGTH_SHORT,
+                            isDisplayedWithBrowserToolbar = false
+                        ).setText("Password cannot be blank.").show()
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:112]: I think this is supposed to show an error state that says "Password required"? We should probably also update that error state as the user is typing and not just when they submit. If we still want to show the snackbar, it also needs to be pulled into strings?

File: app/src/main/res/color/toggle_save_enabled.xml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:color="?disabled" android:state_checked="true" />
```
Comment (ekager) [inline app/src/main/res/color/toggle_save_enabled.xml:6]: should this be state_checked `false` for the disabled state? 🤔

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        return view
+    }
+
+    @ObsoleteCoroutinesApi
+    @ExperimentalCoroutinesApi
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        consumeFrom(savedLoginsStore) {
+            loginDetailView.update(it)
+            login = savedLoginsStore.state.currentItem
+//            showToolbar(savedLoginsStore.state.currentItem?.origin ?: "")
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt:70]: nit remove comment

File: app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            true
+        }
+        else -> false
+    }
+
+    // TODO: Move interactions with the component's password storage into a separate datastore
+    // This includes Delete, Update/Edit, Create
+    private fun attemptSaveAndExit() {
+        var saveLoginJob: Deferred<Unit>? = null
+        lifecycleScope.launch(IO) {
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:134]: Think we can use `viewLifecycleOwner.lifecycleScope` here

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+                    R.string.login_copied
+                )
+            )
+        }
+    }
+
+    // TODO: Move interactions with the component's password storage into a separate datastore
+    private fun fetchLoginDetails() {
+        var deferredLogin: Deferred<List<Login>>? = null
+        val fetchLoginJob = lifecycleScope.launch(Dispatchers.IO) {
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt:110]: nit - use `viewLifecycleOwner.lifecycleScope`

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                create()
+            }.show()
+        }
+    }
+
+    // TODO: Move interactions with the component's password storage into a separate datastore
+    // This includes Delete, Update/Edit, Create
+    private fun deleteLogin() {
+        var deleteLoginJob: Deferred<Boolean>? = null
+        val deleteJob = lifecycleScope.launch(IO) {
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt:178]: same nit - use `viewLifecycleOwner.lifecycleScope`

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            Pair(copyWebAddress, login?.origin),
+            Pair(copyUsername, login?.username),
+            Pair(copyPassword, login?.password)
+        )
+
+        for (copyButtons in copyInfoArgs) {
+            copyButtons.first.setOnClickListener(
+                CopyButtonListener(
+                    copyButtons.second,
+                    R.string.login_copied
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt:101]: These all are supposed to have the same message displayed when you copy them? Shouldn't it be specific to what has been copied? 🤔

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+            true
+        }
+        else -> false
+    }
+
+    private fun editLogin() {
+        val directions =
+            LoginDetailFragmentDirections
+                .actionLoginDetailFragmentToEditLoginFragment(login!!)
+        requireView().findNavController().navigate(directions)
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt:155]: nit - I don't think you need the `requireView` here, `findNavController` should work on the fragment

**Comment by ekager** [comment]
Ah I see now that error states are broken out into a different ticket so disregard those comments. The logins dialog fragment that lives in AC may help with those and also we can compare the password reveal/hide logic we're using there! https://github.com/mozilla-mobile/android-components/blob/master/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/LoginDialogFragment.kt#L185

File: app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
+        inflater.inflate(R.menu.login_save, menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
+        R.id.save_login_button -> {
+            closeKeyboard()
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:102]: There's already an extension method for this on `View`. `view?.hideKeyboard()` should work!

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+//            showToolbar(savedLoginsStore.state.currentItem?.origin ?: "")
+            setUpCopyButtons()
+            showToolbar(savedLoginsStore.state.currentItem?.origin ?: "")
+            setUpPasswordReveal()
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        setHasOptionsMenu(true)
+        super.onCreate(savedInstanceState)
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginDetailFragment.kt:79]: nit - let's call the super lifecycle method first

File: app/src/main/java/org/mozilla/fenix/Presenter.kt [hunk 16]
```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/Presenter.kt:5]: Can we get rid of this file and the `SavedLoginsPresenter`? I don't see how they're used and I'd rather not add new complexity to our current arch if we can get around it

File: app/src/main/res/menu/login_delete.xml (hunk 4 above)
Comment (eliserichards) [reply app/src/main/res/menu/login_delete.xml:7]: No, with the feature flag we want to keep this

File: app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -1197,12 +1197,14 @@
...
+    <!-- Shown in snackbar to tell user that an item has been copied -->
+    <string name="login_copied">Copied to clipboard</string>
     <!-- Shown in snackbar to tell user that the password has been copied -->
     <string name="logins_password_copied">Password copied to clipboard</string>
     <!-- Shown in snackbar to tell user that the username has been copied -->
     <string name="logins_username_copied">Username copied to clipboard</string>
-    <!-- Shown in snackbar to tell user that the site has been copied -->
-    <string name="logins_site_copied">Site copied to clipboard</string>
+    <!-- Shown in snackbar to tell user that the web address has been copied -->
+    <string name="logins_web_address_copied">Site copied to clipboard</string>
```
Comment (ekager) [inline app/src/main/res/values/strings.xml:1207]: Can we revert this change? I'm not sure if we can change the IDs of strings without re-translating them but we can also check with l10n :/

Comment (eliserichards) [reply app/src/main/res/values/strings.xml:1207]: good point

**Comment by eliserichards** [comment]
Waiting for #9986 to be merged first so this doesn't cause merge conflicts.

File: app/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -1211,7 +1211,7 @@
     <string name="logins_password_copied">Password copied to clipboard</string>
     <!-- Shown in snackbar to tell user that the username has been copied -->
     <string name="logins_username_copied">Username copied to clipboard</string>
-    <!-- Shown in snackbar to tell user that the site has been copied -->
+    <!-- Shown in snackbar to tell user that the web address has been copied -->
```
Comment (ekager) [inline app/src/main/res/values/strings.xml:1214]: same l10n issue here I believe. I don't think we can change the description after it has been exported :/

File: app/src/main/res/menu/login_delete.xml [hunk 19]
```diff
@@ -1,14 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
+
```
Comment (ekager) [inline app/src/main/res/menu/login_delete.xml:2]: nit - can we remove the whitespace churn if we didn't actually change this file otherwise?

File: app/src/main/res/drawable-v26/ic_menu_kebab.xml [hunk 20]
```diff
@@ -0,0 +1,16 @@
+<!--
```
Comment (ekager) [inline app/src/main/res/drawable-v26/ic_menu_kebab.xml:1]: Can we remove all the kebab files if they are unused now?

**Comment by AmyYLee** [comment]
@eliserichards Is this ready for a UX review?

File: app/src/main/res/menu/login_delete.xml (hunk 19 above)
Comment (eliserichards) [reply app/src/main/res/menu/login_delete.xml:2]: We're renaming it, so I just accepted the whitespaces

**Comment by AmyYLee** [comment]
Hi, 

Here's my UX review and edits for this feature:

1. Site - Hairline shouldn’t appear when you tap on site URL

2. Tapping on Password field should automatically reveal password with “eye” icon updating to reflect the change

3. Show/Hide password icon in “Edit” mode should follow standard icon colour for each theme. The icon colours for “hide/reveal” and “copy” on the login entry screen after the user exits Edit mode should also follow the same icon colour theme as app. It looks like these are semi-transparent. 

4. When tapping on “Edit” to edit login, the layout of elements on the screen shifts slightly. Layout shouldn’t shift.

**Comment by eliserichards** [comment]
Tap to reveal password: 
tap_to_reveal

Tap site doesn't reveal underline: 
tap_site

Edit & detail view consistency:
edit

**Comment by AmyYLee** [comment]
> Tap to reveal password:
> tap_to_reveal
> 
> Tap site doesn't reveal underline:
> tap_site
> 
> Edit & detail view consistency:
> edit
> 
> Icon consistency will be covered in #10173 or #10172

Hi, 

    Site - Hairline shouldn’t appear when you tap on site URL
**Review - Approved**

    Tapping on password entry field to edit should automatically reveal password with "eye" icon updating to reflect the change. Currently tapping on the password to edit doesn't reveal password.

**Review - Approved**

    Show/Hide password icon in “Edit” mode should follow standard icon colour for each theme. The icon colours for “hide/reveal” and “copy” on the login entry screen after the user exits Edit mode should also follow the same icon colour theme as app. It looks like these are semi-transparent.

@eliserichards comment above - Icon consistency will be covered in #10173 or #10172


    When tapping on “Edit” to edit login, the layout of elements on the screen shifts slightly. Layout shouldn’t shift.

Spoke to @eliserichards - Once the error messages are in then can finalize the positions

File: app/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -1336,8 +1336,25 @@
...
+    <string name="saved_login_username_description">The editable text field for the username of the login.</string>
+    <!--  The editable text field for a login's password. -->
+    <string name="saved_login_password_description">The editable text field for the password of the login.</string>
+    <!--  The button description to save changes to an edited login. -->
+    <string name="save_changes_to_login">Save changes to login.</string>
+    <!--  The button description to discard changes to an edited login. -->
+    <string name="discard_changes">Discard changes</string>
+    <!--  The page title for editing a saved login. -->
+    <string name="edit">Edit</string>
+    <string name="saved_login_password_required">Password required</string>
```
Comment (ekager) [inline app/src/main/res/values/strings.xml:1359]: nit - this new string needs a description for translation

File: app/src/main/res/menu/login_options_menu.xml [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/loginOptionsEditDelete"
+    android:contentDescription="@string/login_options_menu"
+    android:title="@string/login_options_menu"
+    app:showAsAction="never" >
+
+    <item
+        android:id="@+id/edit_login_button"
+        android:contentDescription="@string/login_menu_delete_button"
```
Comment (ekager) [inline app/src/main/res/menu/login_options_menu.xml:16]: wrong content description here?

**Comment by vrcca** [comment]
🚀
